### PR TITLE
Validate Socket.io received "data" messages

### PR DESCRIPTION
**Pull request by ivanslabbert**
Add DTO classes used by class-transformer and class-validator and implement for received Socket.io "data" messages

### CODE REVIEW

File: api/src/permissions/permissions.service.spec.ts [hunk 1]
```diff
@@ -84,17 +84,17 @@ describe("PermissionService", () => {
         const res = Group.getAccess(
             ["group-super-admins"],
             [
-                DocType.Audio,
+                // DocType.Audio,
```
Comment (ivanslabbert) [inline api/src/permissions/permissions.service.spec.ts:87]: Temporarily commented out until DTO's are added for media documents.

File: api/src/dto/ChangeDto.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+export class ChangeDto {
+    @IsNotEmpty()
+    _id: Uuid;
+    @IsNotEmpty()
+    type: DocType.Change;
+    @IsNotEmpty()
+    docId: Uuid;
+    @IsNotEmpty()
+    docType: DocType;
+    memberOf?: Uuid[];
```
Comment (laurensgroeneveld) [inline api/src/dto/ChangeDto.ts:17]: How would we handle advanced validation rules? Just thinking that here we specifically want to check if the passed Uuid's are valid group id's and if the user has access to them. Or do you think we should handle that in the API code and just discard any non-valid id's?

File: api/src/dto/ChangeDto.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import { GroupAclEntryDto } from "./GroupAclEntryDto";
+
+/**
+ * Database structured Post object
+ */
+export class ChangeDto {
+    @IsNotEmpty()
+    _id: Uuid;
+    @IsNotEmpty()
+    type: DocType.Change;
```
Comment (laurensgroeneveld) [inline api/src/dto/ChangeDto.ts:12]: Does this automatically check if the type is "change"?

File: api/src/dto/ChangeReqAckDto.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+/**
+ * API acknowledgement to client after client submission of an ChangeReqDto document.
+ */
+export class ChangeReqAckDto {
+    @IsNotEmpty()
+    docId: Uuid; // Uuid of submitted ChangeReqDto to be acknowledged
+    @IsNotEmpty()
+    type: DocType.ChangeReqAck;
+    @IsNotEmpty()
+    ack: AckStatus;
```
Comment (laurensgroeneveld) [inline api/src/dto/ChangeReqAckDto.ts:13]: Does this validate that it is one of the allowed enum values?

File: api/src/dto/ContentDto.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    @IsString()
+    title: string;
+    @IsString()
+    summary: string;
+    @IsString()
+    author: string;
+    @IsString()
+    text: string;
+    @IsString()
+    seo: string;
```
Comment (laurensgroeneveld) [inline api/src/dto/ContentDto.ts:29]: Small note, I propose to leave this out for now as I don't think "seo" will actually be a string, more likely a separate class with several fields such as description and keywords. This is what the current site has:

image

I left it out in the frontend for now for that reason as well.

**Review by laurensgroeneveld — COMMENTED** [review]
I just have some small questions on the validation library and how we handle more advanced cases.

In general I think we need some tests for this validation, or do you plan on adding those in the follow-up PR with validation messages and errors to the client? Which would be fine with me.

File: api/src/dto/ChangeReqAckDto.ts (hunk 4 above)
Comment (ivanslabbert) [reply api/src/dto/ChangeReqAckDto.ts:13]: Added an @Equals decorator to test for a specific value

File: api/src/dto/ChangeDto.ts (hunk 3 above)
Comment (ivanslabbert) [reply api/src/dto/ChangeDto.ts:12]: Adding a Equals decorator to ensure the correct value

File: api/src/dto/ChangeDto.ts (hunk 2 above)
Comment (ivanslabbert) [reply api/src/dto/ChangeDto.ts:17]: Can add multiple validation decorations. Access validation will be done separately in the API by making calls to the permission system

File: api/src/dto/ContentDto.ts (hunk 5 above)
Comment (ivanslabbert) [reply api/src/dto/ContentDto.ts:29]: Commented out for now

File: api/src/dto/ChangeDto.ts [hunk 6]
```diff
@@ -1,21 +1,39 @@
...
     docType: DocType;
+
+    @IsOptional()
+    @IsArray()
+    @IsString({ each: true })
     memberOf?: Uuid[];
+
+    @IsOptional()
+    @ValidateNested({ each: true })
+    @Type(() => GroupAclEntryDto) // This throws an exception on validation failure, so we need to catch the error on validation. The message is less user-friendly but at least the validator fails and will protect our data.
```
Comment (laurensgroeneveld) [inline api/src/dto/ChangeDto.ts:34]: It's possible to customize the message which would make sense if it's not super clear (and especially if we want to display it in the frontend): https://github.com/typestack/class-validator?tab=readme-ov-file#validation-messages

File: api/src/dto/ChangeReqAckDto.ts [hunk 7]
```diff
@@ -1,16 +1,23 @@
...
  * API acknowledgement to client after client submission of an ChangeReqDto document.
  */
 export class ChangeReqAckDto {
     @IsNotEmpty()
-    docId: Uuid; // Uuid of submitted ChangeReqDto to be acknowledged
+    @IsString()
+    reqId: Uuid; // Uuid of submitted ChangeReqDto to be acknowledged
+
     @IsNotEmpty()
+    @Equals("changeReqAck")
```
Comment (laurensgroeneveld) [inline api/src/dto/ChangeReqAckDto.ts:13]: Did you try putting `DocType.ChangeReqAck` here? Then we don't have to use the string value directly.

File: api/src/dto/ChangeReqAckDto.ts [hunk 8]
```diff
@@ -1,16 +1,23 @@
...
+import { Equals, IsEnum, IsNotEmpty, IsOptional, IsString } from "class-validator";
 
 /**
  * API acknowledgement to client after client submission of an ChangeReqDto document.
  */
 export class ChangeReqAckDto {
     @IsNotEmpty()
-    docId: Uuid; // Uuid of submitted ChangeReqDto to be acknowledged
+    @IsString()
+    reqId: Uuid; // Uuid of submitted ChangeReqDto to be acknowledged
```
Comment (laurensgroeneveld) [inline api/src/dto/ChangeReqAckDto.ts:10]: There is also a `@IsUUID` assertion which I think fits more for these cases.

File: api/src/dto/ContentDto.ts [hunk 9]
```diff
@@ -1,36 +1,60 @@
...
     @IsString()
+    @IsNotEmpty()
     title: string;
+
     @IsString()
+    @IsNotEmpty()
     summary: string;
+
     @IsString()
+    @IsOptional()
```
Comment (laurensgroeneveld) [inline api/src/dto/ContentDto.ts:30]: For readability (but perhaps it's a hassle to change it now) I would propose to put `IsOptional` at the top, so it's easier to spot when a field it optional.

File: api/src/socketio.ts [hunk 10]
```diff
@@ -86,26 +110,45 @@ export class Socketio {
         const changeReq = plainToInstance(ChangeReqDto, data);
```
Comment (laurensgroeneveld) [inline api/src/socketio.ts:110]: As a general not this file is growing quite a bit now so maybe in a follow-up PR some parts could be extracted, like the validation, which might make testing those specific parts easier as well as they don't rely on socket.io.

**Review by laurensgroeneveld — APPROVED** [review]
Some small comments but I'll approve because the rest looks good 👍 Would like some more test coverage for the validation of added documents but that can also be handled in another PR with the permission check for example.

File: api/src/dto/ChangeReqAckDto.ts (hunk 8 above)
Comment (ivanslabbert) [reply api/src/dto/ChangeReqAckDto.ts:10]: I left it as @IsString to allow using our current test data set. Might change it later though...

Comment (ivanslabbert) [reply api/src/dto/ChangeReqAckDto.ts:10]: I left it as @IsString to allow using our current test data set. Might change it later though...

Comment (ivanslabbert) [reply api/src/dto/ChangeReqAckDto.ts:10]: I left it as @IsString to allow using our current test data set. Might change it later though...

File: api/src/dto/ContentDto.ts (hunk 9 above)
Comment (ivanslabbert) [reply api/src/dto/ContentDto.ts:30]: Agree

File: api/src/dto/ChangeReqAckDto.ts (hunk 7 above)
Comment (ivanslabbert) [reply api/src/dto/ChangeReqAckDto.ts:13]: Seems to also work.

File: api/src/dto/ChangeDto.ts (hunk 6 above)
Comment (ivanslabbert) [reply api/src/dto/ChangeDto.ts:34]: Throwing an exception makes it not show the message at all. So not very helpful here. Custom messages are however nice, and can be used when we add more detailed validations.

**Pull request merged**